source/encoding.c

```diff
@@ -0,0 +1,316 @@
+
+    size_t encoded_length = calculate_base64_encoded_length(to_encode_len);
+    if(!output) {
+        *output_size = encoded_length;
+        return AWS_OP_SUCCESS;
+    }
+
+    if(*output_size < encoded_length) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    *output_size = encoded_length;
+
+    size_t buffer_length = to_encode_len;
+    size_t block_count = (buffer_length + 2) / 3;
+    size_t remainder_count = (buffer_length % 3);
+    size_t str_index = 0;
+
+    for(size_t i = 0; i < to_encode_len; i += 3 )
+    {
+        uint32_t block = to_encode[ i ];
+
+        block <<= 8;
+        if (i + 1 < buffer_length)
+        {
+            block = block | to_encode[ i + 1 ];
+        }
+
+        block <<= 8;
+        if (i + 2 < to_encode_len)
+        {
+            block = block | to_encode[ i + 2 ];
+        }
+
+        output[str_index++] = base64_encoding_table[(block >> 18) & 0x3F];
+        output[str_index++] = base64_encoding_table[(block >> 12) & 0x3F];
+        output[str_index++] = base64_encoding_table[(block >> 6) & 0x3F];
+        output[str_index++] = base64_encoding_table[block & 0x3F];
+    }
+
+    if(remainder_count > 0)
+    {
+        output[block_count * 4 - 1] = '=';
+        if(remainder_count == 1)
+        {
+            output[block_count * 4 - 2] = '=';
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static void init_decoding_table() {
+    if(!base64_decoding_table_init) {
+        for(size_t i = 0; i < (sizeof(base64_encoding_table) / sizeof(char)); ++i)
+        {
+            size_t index = (size_t)base64_encoding_table[i];
+            base64_decoding_table[index] = (uint8_t)i;
+        }
+
+        base64_decoding_table[(size_t)'='] = base64_sentinal_value;
+        base64_decoding_table_init = 1;
+    }
+}
+
+int aws_base64_decode(const char *to_decode, size_t to_decode_len, uint8_t *output, size_t *output_size) {
+    size_t decoded_length = calculate_base64_decoded_length(to_decode, to_decode_len);
+
+    if(!output) {
+        *output_size = decoded_length;
+        return AWS_OP_SUCCESS;
+    }
+
+    init_decoding_table();
+
+    if (*output_size < decoded_length) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    *output_size = decoded_length;
+
+    size_t block_count = to_decode_len / 4;
+    size_t string_index = 0;
+
+    for(size_t i = 0; i < block_count; ++i)
+    {
+        uint32_t value1 = base64_decoding_table[(size_t)to_decode[string_index++]];
+        uint32_t value2 = base64_decoding_table[(size_t)to_decode[string_index++]];
+        uint32_t value3 = base64_decoding_table[(size_t)to_decode[string_index++]];
+        uint32_t value4 = base64_decoding_table[(size_t)to_decode[string_index++]];
+
+        size_t buffer_index = i * 3;
+        output[buffer_index++] = (uint8_t)((value1 << 2) | ((value2 >> 4) & 0x03));
+        if(value3 != base64_sentinal_value)
```

Comment:
The main loop shouldn't be making all these tests; we should make the loop go for `block_count - 1` iterations and do all these tests at the final iteration only.